### ISSUE TITLE
Salt Cloud.CloudClient has a breaking typo,

### ISSUE THREAD

**Issue by fake-name** [saltstack/salt#41971]
### Description of Issue/Question
Trying to use a instance of `salt.cloud.CloudClient('/etc/salt/cloud')`, you cannot create any VPSes, because apparently someone got halfway through refactoring `salt/salt/cloud/__init__.py`, and then committed it. 

Basically, I do `cloudclient.create(names=[clientname], provider=provider, **kwargs)`. This fails:

```
Traceback (most recent call last):
  File "salt_runner.py", line 147, in make_client
    ret = self.cc.create(names=[clientname], provider=provider, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 404, in create
    mapper.create(vm_))
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1182, in create
    alias, driver = vm_['provider'].split(':')
KeyError: 'provider'
```

Attaching with pdb:

```
> /usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py(1176)create()
-> output = {}
(Pdb) locals
<built-in function locals>
(Pdb) locals()
{'vm_': {'profile': None, 'name': 'scrape-worker-4', 'image': 'ubuntu-14-04-x64', 'driver': 'do:digital_ocean', 'location': 'nyc3', 'private_networking': False, 'vm_size': '512mb', 'size': '512mb'}, 'local_master': True, 'self': <salt.cloud.Map object at 0x70fa9dc75090>}
(Pdb) bt
  /usr/lib/python2.7/bdb.py(400)run()
-> exec cmd in globals, locals
  <string>(1)<module>()
  /home/client/AutoTriever/marshaller/salt_scheduler.py(139)<module>()
-> sched.ensure_active_workers()
  /home/client/AutoTriever/marshaller/salt_scheduler.py(95)ensure_active_workers()
-> self.create_vm(vm_name)
  /home/client/AutoTriever/marshaller/salt_scheduler.py(48)create_vm()
-> self.interface.make_client(vm_name)
  /home/client/AutoTriever/marshaller/salt_runner.py(147)make_client()
-> ret = self.cc.create(names=[clientname], provider=provider, **kwargs)
  /usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py(404)create()
-> mapper.create(vm_))
> /usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py(1176)create()
-> output = {}

```

So the issue is that apparently `salt/cloud/__init__.py(1176)create()` is expecting a key named `provider`, but the parameter is being inserted into the parameter dictionary by `salt/cloud/__init__.py(404)create()` with the key `driver`. It seems like someone was intending to rename the parameter key, but only did half of it.

I guess that's what I get for trying to update.

**Comment by fake-name** [comment]
Wait, working through past commits, apparently the `provider` parameter was depreciated at some point.

88936c2c84ba08b041c5b9f31a03944bf22bf1f4 claims that there are depreciation warnings, but I never saw them. I think there might be an assumption that the only way people use salt cloud is via yaml files? My entire orchistration system is done from python, and the depreciation warnings were never visible when using a CloudClient() directly.

**Comment by fake-name** [comment]
So, somewhere between 2016.11, I somehow completely lost the ability to create VPSes. `CloudClient().create()` from `develop`  didn't throw any error, but it also did nothing. 

I may have a change to poke about ant try to figure out what is going on this weekend, but I can't leave my system broken for too long, I don't have a salt testing environment, unfortunately (it requires a routable public IP, and I don't have many of those to spare).

**Comment by rallytime** [comment]
Hi @fake-name - I am looking into this issue right now to see where we're not handling your use-case. Can you tell me what version of salt you upgraded from where this was working? And you were trying out the `develop` branch, correct?

**Comment by fake-name** [comment]
I switched from somewhere between `2016.3` and `2016.11` to `develop`, and things broke. I rolled back to `2016.11`, and things started working again.

It's probably worth noting for a lot of this that I'm *probably* not necessarily using salt the way most of the documentation seems to assume people will. I'm not interested in running saltstack as a stand-alone tool, but rather using it as a component in a larger python script. As such, I'm not using the CLI interface or the salt state config file thing at all. 

My interface script thing is here: https://github.com/fake-name/AutoTriever/blob/master/marshaller/salt_runner.py

Basically, I've had to more or less guess and experiment to figure out how to use the python interface components, and as such, I'm *probably* doing things incorrectly, or at least making bad assumptions.

**Comment by rallytime** [comment]
@fake-name Ah, yeah. The deprecation warnings have been in the code since the 2015.8 releases, but were removed in the `2017.7` and `develop` branches with the commit you referenced above since the deprecation notification time has expired. I am not sure quite yet why you weren't seeing the warnings. I have an idea, which is the same reason why you're seeing the stack trace, but I'd like to confirm something first.

Can you post a *sanitized* version of your `/etc/salt/cloud` file?

**Comment by rallytime** [comment]
Actually, never mind. I figured out where the issue is. Now I just need to figure out the best way to fix it. I'll submit something by the end of the day. Thanks for filing this issue. You're right that using the `create` function with the CloudClient in this case is not being handled the right way. I'll add some tests for this as well.

**Comment by fake-name** [comment]
I found the depreciation warnings at one point, but I'll be damned if I can remember where. Basically, AFICT, the issue is the warning is thrown when you load a config file, rather then when you use the depreciated function. This fails in my case because I have no configuration file, as I'm interacting with the API directly. 

My `/etc/salt/cloud` file has no changes from the default, I've literally never edited it. the only changes from a bare install I've made is by adding keys to `/etc/salt/cloud.providers`, as far as I can tell (I might have set the master hostname at some point, too, I'm not sure.

Everything else is done dynamically, as I've said. The script that does all this is https://github.com/fake-name/AutoTriever/blob/master/marshaller/salt_runner.py.

**Comment by rallytime** [comment]
Thanks for the extra information @fake-name. The custom runner you posted was helpful.

Yes, the warning is thrown when you load the config file. I am having a hard time understanding how you are loading the keys in `/etc/salt/cloud.providers`. Because when that file is loaded, the warning should have notified you. It wasn't a function that was deprecated, it was the syntax used with salt-cloud provider files that was deprecated. The word `provider` is heavily overloaded in salt-cloud, so using `driver` was an attempt to reduce a lot of confusion when setting up salt-cloud config files (whether you're using the CloudClient or not). In the 2017.7 and develop branches of salt, you _must_ have `driver` defined somewhere in your provider configuration.

Anyway, there is definitely a bug here in the `2017.7` and `develop` branches when using the CloudClient create function. I have reproduced the stacktrace you're hitting and I have a fix with tests incoming sometime later today. Thanks for the help in debugging this and for filing the issue.

**Comment by fake-name** [comment]
Ohhhhhhh, derp derp derp derp. I am using `driver` in my cloud.providers file. 

```
client@rabbit:~$ cat /etc/salt/cloud.providers
do:
  driver: digital_ocean
  personal_access_token: <snip>
  ssh_key_file: /etc/salt/pki/cloud/id_rsa
  ssh_key_names: DesktopKey,SaltKey,HerpKey
  create_dns_record: False

  script: bootstrap-salt
  minion:
    master: <snip>

vultr:
  # Vultr account api key
  api_key: <snip>
  driver: vultr
  ssh_key_file: /etc/salt/pki/cloud/id_rsa
  ssh_key_names: DesktopKey,SaltKey,HerpKey
  script: bootstrap-salt

  script: bootstrap-salt
  minion:
    master: <snip>

linode:
  apikey: '<snip>'
  password: '<snip>'
  driver: linode

  script: bootstrap-salt
  minion:
    master: <snip>

scaleway:
  access_key: <snip>
  token: <snip>
  driver: scaleway

  script: bootstrap-salt
  minion:
    master: <snip>
```

I was passing the driver selection as a parameter named `provider` when interacting with the API, though.

**Comment by rallytime** [comment]
Ah! Great! I was trying to figure out where that relationship with the cloud driver files was happening. The way you're doing it here is correct, but there is definitely a bug there. I have fixed this in #42034 and added an integration test so we can catch these kinds of breaks in the future.

Thanks again @fake-name!

**Comment by fake-name** [comment]
Ok, assuming salt behaves in a virtualenv, I can confirm that it appears I can successfully create VMs with the `develop` branch. 

I'm running in a virtualenv into which I installed salt, on a computer with an older version of salt outside of the virtualenv. I'm not sure if this could confound my testing, since I think salt has a daemon, so I'm not sure if I'm accidentally talking to the older version or not.
